subsys/net/lib/nrf_cloud/src/nrf_cloud_log_backend.c

```diff
@@ -394,31 +457,47 @@ static int send_ring_buffer(void)
 	if (!output.data.len) {
 		goto cleanup;
 	}
-
-	uint8_t *p = (uint8_t *)output.data.ptr;
-
-	p[output.data.len] = '\0';
+	if ((log_format_current == LOG_OUTPUT_DICT) &&
+	    IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		topic = get_dict_topic();
+		if (!topic) {
+			goto cleanup;
+		}
+		err = convert_to_quoted_base64(output.data.ptr, output.data.len,
+					       (uint8_t **)&output.data.ptr, &output.data.len);
+		if (err) {
+			goto cleanup;
+		}
+		LOG_DBG("topic:   %s", topic);
+		LOG_DBG("payload: %s", (const char *)output.data.ptr);
+	} else {
+		/* Send ring buffer contents as is -- JSON or raw binary. */
+		((uint8_t *)output.data.ptr)[output.data.len] = '\0';
+	}
 
 	LOG_DBG("Ready to transmit %zd bytes...", output.data.len);
 	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
 		err = nrf_cloud_send(&output);
 	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
 		do {
+			/* If no topic provided, use d2c/bulk, otherwise use topic. */
 			err = nrf_cloud_rest_send_device_message(log_context.rest_ctx,
 								 log_context.device_id,
-								 output.data.ptr, true, NULL);
+								 output.data.ptr,
+								 !topic, topic);
 			if (err) {
 				LOG_ERR("Error sending message:%d", err);
-				if (p[0] == '\0') {
-					LOG_ERR("Empty buffer!");
-				}
 				if (err != -EBUSY) {
 					LOG_ERR("Data: %s, len: %zd",
 						(const char *)output.data.ptr, output.data.len);
 				}
 				k_sleep(K_MSEC(100));
 			}
 		} while (err == -EBUSY);
+		if (log_format_current == LOG_OUTPUT_DICT) {
+			nrf_cloud_free((void *)output.data.ptr);
```

Comment:
I think instead of using `output.data.ptr` over and over in these various steps, we should set up separate pointers for the different ways in which this data can get handled.

It's weird to have this conditional `nrf_cloud_free` on a pointer that absolutely should not be freed if we are in non-REST mode

Reply:
So, instead of 

```c
output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb, (uint8_t **)&output.data.ptr, stored);
```

I suggest 
```c
uint8_t *log_rb_chunk;
int log_rb_chunk_len = ring_buf_get_claim(&log_nrf_cloud_rb, &log_rb_chunk, stored);
```

Then, inside the REST sections, have

```c
uint8_t *log_b64;
int log_b64_len;
err = convert_to_quoted_base64(log_rb_chunk, log_rb_chunk_len, &log_b64, &log_b64_len);
```

Then during cleanup you can just have

```c
if (log_b64) {
    nrf_cloud_free(log_b64);
}
```

Then, do a conditional population of the output struct, looking something like this

```c
if (REST dict logs) {
	output.data.len = log_b64_len;
	output.data.buf = log_b64;
} else {
	output.data.len = log_rb_chunk_len;
	output.data.buf = log_rb_chunk;
}
```

Exact names of these variables being up to your choice.

I think this would be more readable, and easier to understand/maintain